File: webapp/backend/schemas/factory.py
"""
工厂配置相关数据模型
"""
from pydantic import BaseModel, Field
from typing import Dict, List, Optional


class Route(BaseModel):
    from_: str = Field(..., alias="from")
    to: str
    material: str
    batch_size: int
    travel_time: float = 0
    transporter_id: str

    class Config:
        populate_by_name = True


class Assembly(BaseModel):
    station: str
    inputs: Dict[str, int]
    output: str
    process_time: float


class MonitorItem(BaseModel):
    node: str
    material: str


class Transporter(BaseModel):
    id: str
    count: int
    speed: float


class Summary(BaseModel):
    finished_node: str
    finished_material: str


class FactoryConfig(BaseModel):
    initial_inventory: Dict[str, Dict[str, int]]
    routes: List[Route]
    assemblies: List[Assembly]
    summary: Summary
    monitor: List[MonitorItem]
    transporters: List[Transporter]
    layout: str


class FunctionalUnit(BaseModel):
    id: str
    label: str
    width: float
    height: float
    movable: bool = True
    x: Optional[float] = None
    y: Optional[float] = None
    angle: Optional[int] = 0


class Obstacle(BaseModel):
    id: str
    label: str
    width: float
    height: float
    movable: bool = False
    x: Optional[float] = None
    y: Optional[float] = None
    angle: Optional[int] = 0


class Canvas(BaseModel):
    width: float
    height: float


class LayoutConfig(BaseModel):
    canvas: Canvas
    fus: List[FunctionalUnit]
    obstacles: List[Obstacle] = []

File: webapp/frontend/src/types/factory.ts
// Core data models for factory layout and logistics planning

export type MaterialId = string;
export type NodeId = string;
export type TransporterId = string;
export type RouteId = string;

export interface Material {
  id: MaterialId;
  color?: string;
  description?: string;
}

export interface TransporterDefinition {
  id: TransporterId;
  speed: number;
  count: number;
}

export type FactoryNodeType = "FU" | "Obstacle";

export interface FactoryNode {
  id: NodeId;
  type: FactoryNodeType;
  dimensions: {
    length: number;
    width: number;
    notchLength?: number;
    notchWidth?: number;
  };
  angle?: number;
  x?: number;
  y?: number;
  initialInventory?: Record<MaterialId, number>;
}

export interface Assembly {
  nodeId: NodeId;
  inputs: Record<MaterialId, number>;
  output: MaterialId;
  processTime: number;
  outputCount?: number; // 默认1，可选
}

export interface Route {
  id: RouteId;
  fromNodeId: NodeId;
  toNodeId: NodeId;
  materialId: MaterialId;
}

export interface RouteConfiguration {
  routeId: RouteId;
  transporterId: TransporterId;
  batchSize: number;
  travelTime: number;
}

// Aggregate state for convenience (used in store)
export interface FactoryStateSnapshot {
  materials: Material[];
  transporters: TransporterDefinition[];
  nodes: FactoryNode[];
  assemblies: Assembly[];
  routes: Route[];
  routeConfigs: RouteConfiguration[];
  canvasSize: { width: number; height: number };
}
